### DISCUSSION TITLE
How does FarNet work in terms of pairing managed code with FAR?

### DISCUSSION THREAD

**Discussion by Usurer** [nightroman/FarNet#39] [Q&A] [answered]
Hi! 
I've got a question about the basics of FarNet operation.
As far as I understand, FAR is written in C++, whereas FarNet is a managed .NET code. I can't figure out how this integration is implemented, i.e. **how does FAR recognize FarNet.dll as a valid plugin**?
The reason I'm asking isn't just sheer curiosity, but I was looking for a way to understand how does FAR API (as described at https://api.farmanager.com/) correspond to FarNet API.
I'd greatly appreciate you pointing me in the right direction - docs, keywords to google, source code to read - anything )

**Comment by nightroman** [comment, marked as answer]
This is a good question, interesting for those who want to connect C++ and .NET.

> How does FAR recognize FarNet.dll as a valid plugin?

It does not. FAR knows nothing about FarNet.dll.
FarNet.dll is a pure .NET assembly with API and tools for .NET modules and scripts.
API is mostly interfaces and abstract classes, not implemented in FarNet.dll itself.

But FarNet.dll is not the FarNet engine. The engine is Plugins\FarNet\FarNetMan.dll.
FarNetMan.dll is built with C++/CLI, i.e. it is native and managed at the same time.
Thus:

- FAR treats FarNetMan.dll as the usual plugin (it's possible because FarNetMan.dll is native)
- FarNetMan.dll loads FarNet.dll and connects to implemented instances (it's possible because FarNetMan.dll is also managed)
- Then FarNet modules are discovered in FarNet\Modules and loaded, some are cached, some are always loaded (PowerShellFar)

> to understand how does FAR API (as described at https://api.farmanager.com/) correspond to FarNet API

FAR API corresponds to FarNet API of course, in a way, but not literally at all.
FarNet API is designed as object oriented and .NET friendly, with PowerShell in mind, too.
If you are about to work on FarNet modules and scripts then you do not need https://api.farmanager.com/, well, almost.

**Reply by nightroman** [reply]
The source code is https://github.com/nightroman/FarNet/tree/master/FarNet
The folder FarNet -- FarNet API and tools, written in C#.
The folder FarNetMan -- FAR plugin and FarNet loader, written in C++/CLI.

**Reply by Usurer** [reply]
Ah-hah, got it. Thanks!
